apps/src/lib/node/ledger/shell/finalize_block.rs

```diff
@@ -809,6 +838,100 @@ where
 
         Ok(())
     }
+
+    /// Charge fee for the provided wrapper transaction
+    fn charge_fee(
+        &mut self,
+        wrapper: &WrapperTx,
+        gas_table: &BTreeMap<String, u64>,
+        #[cfg(not(feature = "mainnet"))] has_valid_pow: bool,
+    ) -> Result<()> {
+        // Unshield funds if requested
+        if let Some(unshield) = &wrapper.unshield {
+            // The unshielding tx does not charge gas, instantiate a limitless
+            // custom gas meter for this step
+            let mut gas_meter = TxGasMeter::new(u64::MAX);
+
+            // NOTE: given the checks in process_proposal, a failure in the
+            // unshielding tx should not happen. If it does, do not return early
+            // from this function but try to take the funds from the unshielded
+            // balance
+            match apply_tx(
+                TxType::Decrypted(DecryptedTx::Decrypted {
```

Comment:
In here, I fake a decrypted tx to execute the unshielding. It's basically the same trick we use to execute governance proposals. At this point, I don't know whether we should think about allowing `Raw` txs execution (making sure that no other Raw txs can end up in a block and be executed)